python_coreml_stable_diffusion/activation_quantization.py

```diff
@@ -0,0 +1,489 @@
+    unet_cls = unet.UNet2DConditionModel
+    reference_unet = unet_cls(**pipe.unet.config).eval()
+    reference_unet.load_state_dict(pipe.unet.state_dict())
+
+    # Initialize reference pipeline
+    ref_pipe, _ = prepare_pipe(pipe, reference_unet)
+
+    del pipe
+    gc.collect()
+    return ref_pipe
+
+def main(args):
+    # Initialize reference pipeline
+    ref_pipe = get_reference_pipeline(args.model_version)
+    if torch.cuda.is_available():
+        device = "cuda"
+    else:
+        device = "cpu"
+    logger.debug(f"Placing pipe in {device}")
+    ref_pipe.to(device)
+    # Generate baseline outputs
+    ref_out = run_pipe(ref_pipe)
+
+    # Setup artifact file paths
+    os.makedirs(args.o, exist_ok=True)
+    recipe_json_path = os.path.join(args.o, f"{args.model_version.replace('/', '_')}_quantization_recipe.json")
+    calibration_dir = os.path.join(args.o, f"calibration_data_{args.model_version.replace('/', '_')}")
+
+    # Generate calibration data 
+    if args.generate_calibration_data:
+        generate_calibration_data(ref_pipe, args, calibration_dir)
+
+    # Compute layer-wise PSNR
+    if args.layerwise_sensitivity:
+        logger.info("Compute Layer-wise PSNR")
+        quantizable_modules = get_quantizable_modules(ref_pipe.unet)
+
+        results = {
+            'conv': {},
+            'einsum': {},
+            'model_version': args.model_version
+        }
+        dataloader = unet_data_loader(calibration_dir, device, args.calibration_nsamples)
+
+        for module_type, module_name in quantizable_modules:
+            logger.info(f"Quantizing UNet Layer: {module_name}")
+            config = quantize_module_config(module_name)
+            quantized_unet = quantize(ref_pipe.unet, config, dataloader)
+
+            # Generate outputs from quantized model
+            q_pipe, _ = prepare_pipe(ref_pipe, quantized_unet)
+            test_out = run_pipe(q_pipe)
+
+            psnr = [float(f"{compute_psnr(r, t):.1f}") for r, t in zip(ref_out, test_out)]
+            logger.info(f"PSNR: {psnr}")
+            avg_psnr = sum(psnr) / len(psnr)
+            logger.info(f"AVG PSNR: {avg_psnr}")
+            results[module_type][module_name] = avg_psnr
+
+            del quantized_unet
+            del q_pipe
+            gc.collect()
+
+        with open(recipe_json_path, 'w') as f:
+            json.dump(results, f, indent=2)
+
+    if args.quantize_pytorch:
+        logger.info("Quantizing UNet PyTorch model")
+        dataloader = unet_data_loader(calibration_dir, device, args.calibration_nsamples)
+
+        with open(recipe_json_path, "r") as f:
+            results = json.load(f)
+
+        logger.info(f"Conv PSNR threshold: {args.conv_psnr}, Attn PSNR threshold: {args.attn_psnr}")
+        skipped_conv = set([layer for layer, psnr in results['conv'].items() if psnr < args.conv_psnr])
+        skipped_einsum = set([layer for layer, psnr in results['einsum'].items() if psnr < args.attn_psnr])
```

Comment:
I think it might be good to refactor the section from 380 to 389, and the section from 394 to 397 into two separate functions, that override the quantization setting solely based on PSNRs. One method for inference speed up and the other for accuracy. 
Just so that it's easier to experiment for someone by commenting it out as necessary.